WebServices/AzureSignalR/ChatClient/MainPage.xaml.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using Xamarin.Forms;
+
+namespace ChatClient
+{
+    public partial class MainPage : ContentPage
+    {
+        SignalRService signalR;
+
+        public MainPage()
+        {
+            InitializeComponent();
+
+            signalR = new SignalRService();
+            signalR.Connected += SignalR_ConnectionChanged;
+            signalR.ConnectionFailed += SignalR_ConnectionChanged;
+            signalR.NewMessageReceived += SignalR_NewMessageReceived;
+        }
+
+        private void AddMessage(string message)
+        {
+            Device.BeginInvokeOnMainThread(() =>
+           {
```

Comment:
Alignment.

Reply:
Also, just checking, does this have to be marshalled onto the UI thread?

Reply:
^ Looks like the answer is no, as it's invoked by `SignalR_NewMessageReceived`.

Reply:
^ Having said that, I guess the SignalR event handlers may run on background threads?

Reply:
The events are not guaranteed to run on the UI thread, most of them are indeed on background threads. When `AddMessage` is called in `SignalR_ConnectionChanged` it could be called outside of the `BeginInvokeOnMainThread` there, since the `AddMessage` thread does it's own marshalling. However, I left it with the other lines of code for readability. In every other case the UI updates must be forced to UI Thread or the Exception out.